pkg/newstorage/newstorage.go

```diff
@@ -0,0 +1,122 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package newstorage
+
+import "errors"
+
+// ErrStoreNotFound is returned when a store is not found.
+var ErrStoreNotFound = errors.New("store not found")
+
+// ErrDataNotFound is returned when data is not found.
+var ErrDataNotFound = errors.New("data not found")
+
+// StoreConfiguration represents the configuration of a store.
+type StoreConfiguration struct {
+	TagNames []string
+}
+
+// QueryOptions represents various options for Query calls in a store.
+type QueryOptions struct {
+	pageSize int
+}
+
+// QueryOption represents an option for a Query call in a store.
+type QueryOption func(opts *QueryOptions)
+
+// WithPageSize sets the maximum page size for data retrievals done within the Iterator returned by the Query call.
+// Paging is handled internally by the Iterator. Higher values may reduce CPU time and the number of database calls at
+// the expense of higher memory usage.
+func WithPageSize(size int) QueryOption {
+	return func(opts *QueryOptions) {
+		opts.pageSize = size
+	}
+}
+
+// Tag represents a Name + Value pair that can be associated with a k,v pair for querying later.
+type Tag struct {
+	Name  string
+	Value string
+}
+
+// Operation represents an operation to be performed in the Batch method.
+type Operation struct {
+	Key   string
+	Value []byte // A nil value will result in a delete operation
+	Tags  []Tag  // Optional
+}
+
+// Provider represents a storage provider.
+type Provider interface {
+	// OpenStore opens a store with the given name and returns a handle.
+	// If the store has never been opened before, then it is created.
+	OpenStore(name string) (Store, error)
+
+	// SetStoreConfig sets the configuration on a store.
+	// The store must be created prior to calling this method.
+	SetStoreConfig(name string, config StoreConfiguration) error
+
+	// GetStoreConfig gets the current store configuration.
+	// The store must be created prior to calling this method.
+	GetStoreConfig(name string) (StoreConfiguration, error)
+
+	// CloseStore closes the store with the given name.
+	// This does not delete any data in the store, and the store can be reopened again.
+	CloseStore(name string) error
```

Comment:
We might still want to discuss if CloseStore is in the right place.

Reply:
The reason I kept CloseStore in the Provider is because the CouchDB provider implementation (and possibly others) holds on to opened stores as a sort of "cache", so in order to keep that behaviour I had to keep CloseStore in the provider. The other reason I kept it here was for symmetry with the OpenStore method.

Reply:
The idea of a `Close` is to release resources. I'm not really understanding what `CloseStore` is doing in your case.

If resources need to be released, it would make sense to release from the handle. Perhaps I'm missing something?

Reply:
This is what I'm referring to: https://github.com/hyperledger/aries-framework-go-ext/blob/main/component/storage/couchdb/store.go#L119

I feel like the CouchDB provider doesn't really need to hold on to a list of opened store objects. Do you think it's okay to not have that "cache"? (not exactly a cache... more like a shortcut in case you try to open an already existing store)